### PR TITLE
Feature/get build actions

### PR DESCRIPTION
**Pull request by richabindra**
This PR addresses Issue #28.

### CODE REVIEW

File: src/main/java/com/cdancy/jenkins/rest/domain/job/Action.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import org.jclouds.json.SerializedNames;
+
+import java.util.List;
+
+@AutoValue
+public abstract class Action {
+
+    public abstract List<Cause> causes();
+
+    public abstract List<Parameter> parameters();
```
Comment (martinda) [inline src/main/java/com/cdancy/jenkins/rest/domain/job/Action.java]: I wonder if there is a way to make parameters a `Map<String, String>` instead of a plain list, it would make retrieval more natural.

Comment (cdancy) [reply src/main/java/com/cdancy/jenkins/rest/domain/job/Action.java]: There is a way we could marshal it into something else but I'd actually rather not do so and instead hand back to the user the exact response we get from jenkins for better or worse.

Comment (martinda) [reply src/main/java/com/cdancy/jenkins/rest/domain/job/Action.java]: No problem, I am okay with that.

Comment (martinda) [reply src/main/java/com/cdancy/jenkins/rest/domain/job/Action.java]: Maybe I should have dismissed the review... oh well.

Comment (richabindra) [reply src/main/java/com/cdancy/jenkins/rest/domain/job/Action.java]: Yes, I was confused about this too whether to leave it as is or should we convert it to something more useful for the user. But, it makes sense to leave it as is.

Comment (cdancy) [reply src/main/java/com/cdancy/jenkins/rest/domain/job/Action.java]: We're doing this mainly to keep things sane on our end with the least amount of confusion and gotchas. There are definitely good reasons to marshal this into something else but then I don't want to get into the habit of having these one-off cases where the user expects things to be a certain way and then the implementation provides something different. I'd rather hand back to the user, in AutoValue object format, exactly what was given to us by Jenkins.

Comment (richabindra) [reply src/main/java/com/cdancy/jenkins/rest/domain/job/Action.java]: Yes, agreed. So, just clarifying, I don't need to change anything in the PR?

File: src/main/java/com/cdancy/jenkins/rest/domain/job/Cause.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+public abstract class Cause {
+
+    public abstract String shortDescription();
+
+    @Nullable
+    public abstract String userId();
+
+    @Nullable
+    public abstract String userName();
+
```
Comment (martinda) [inline src/main/java/com/cdancy/jenkins/rest/domain/job/Cause.java]: The `src/test/resources/build-info-no-params.json` contains a `_class` item some times. Should it be here?

There is a precedent in [Plugins.java](https://github.com/cdancy/jenkins-rest/blob/master/src/main/java/com/cdancy/jenkins/rest/domain/plugins/Plugins.java#L42).

File: src/test/java/com/cdancy/jenkins/rest/features/JobsApiMockTest.java [hunk 3]
```diff
@@ -536,6 +539,64 @@ public void testBuildJobWithParamsNonExistentJob() throws Exception {
...
+        server.enqueue(new MockResponse().setBody(body).setResponseCode(200));
+        JenkinsApi jenkinsApi = api(server.getUrl("/"));
+        JobsApi api = jenkinsApi.jobsApi();
+        try {
+            List<Parameter> output = api.buildInfo(null,"fish", 10).actions().get(0).parameters();
+            assertNotNull(output);
+            assertTrue(output.get(0).name().equals("bear"));
+            assertTrue(output.get(0).value().equals("true"));
+            assertTrue(output.get(1).name().equals("fish"));
+            assertTrue(output.get(1).value().equals("salmon"));
```
Comment (martinda) [inline src/test/java/com/cdancy/jenkins/rest/features/JobsApiMockTest.java]: Although unlikely to break, there should be a test for an empty parameter value. Maybe Jenkins returns null instead of an empty space.

Comment (cdancy) [reply src/test/java/com/cdancy/jenkins/rest/features/JobsApiMockTest.java]: Agreed to all of the above.

File: src/main/java/com/cdancy/jenkins/rest/domain/job/Action.java (hunk 1 above)
Comment (cdancy) [reply src/main/java/com/cdancy/jenkins/rest/domain/job/Action.java]: Nothing on this end though please do address @martinda comments below and then we can merge and kick a new patch release.

File: src/test/java/com/cdancy/jenkins/rest/features/JobsApiMockTest.java [hunk 4]
```diff
@@ -577,6 +578,37 @@ public void testGetParamsWhenNoBuildParams() throws Exception {
...
+        JobsApi api = jenkinsApi.jobsApi();
+        try {
+            List<Action> output = api.buildInfo(null,"fish", 10).actions();
+            List<Parameter> parameters = new ArrayList<>();
+            for (Action action : output) {
+                if (!action.parameters().isEmpty()) {
+                    parameters = action.parameters();
+                }
+            }
+            for (Parameter parameter : parameters) {
```
Comment (richabindra) [inline src/test/java/com/cdancy/jenkins/rest/features/JobsApiMockTest.java]: I intentionally added the test like this to provide an alternative to get(0) etc for getting params. I can change it if suggested otherwise.

File: src/main/java/com/cdancy/jenkins/rest/domain/job/Action.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.cdancy.jenkins.rest.domain.job;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+import org.jclouds.javax.annotation.Nullable;
```
Comment (martinda) [inline src/main/java/com/cdancy/jenkins/rest/domain/job/Action.java]: `Nullable` is not used in this file.

File: src/test/java/com/cdancy/jenkins/rest/features/JobsApiMockTest.java (hunk 4 above)
Comment (martinda) [reply src/test/java/com/cdancy/jenkins/rest/features/JobsApiMockTest.java]: Well, thanks for the tip, but I don't think it is necessary. In fact it makes the test slightly harder to read.

Comment (cdancy) [reply src/test/java/com/cdancy/jenkins/rest/features/JobsApiMockTest.java]: Agreed. Makes it harder to read. Lets just use what comes OOTB and iterate over that.

**Comment by richabindra** [comment]
I was trying to add the live test to check for null and empty param, but when I have a params map with a key and null value and I try to build the job using this param map, it keeps throwing me an error on line 45 below:

https://github.com/cdancy/jenkins-rest/blob/00cb735066f1df640756cd732df42323c69f039b/src/main/java/com/cdancy/jenkins/rest/binders/BindMapToForm.java#L42-L45

In the above class i.e. BindMapsToForm.java, I locally tried adding a check before line 45 to convert any null value to an empty list.  And that works since I explicitly avoid the null pointer exception during prop.getValue().toArray(...) conversion. 

But, using the POST on the rest api, it is allowing me to pass 'null' value for a key. So, I am not sure should we be converting the 'null' to an empty list in the above class. What do you think?

**Comment by cdancy** [comment]
@richabindra converting the potential null to an empty list sounds like a good idea to me.

**Comment by richabindra** [comment]
@cdancy I tried posting to jenkins using CURL which allowed null values but jenkins-rest does not seem to allow null values to a parameter.
By converting null to an empty list, are we not modifying the existing functionality?

**Comment by martinda** [comment]
@richabindra What does the curl POST look like? How are you POSTing a null value? It is an empty string, is is the string "null"?

**Comment by richabindra** [comment]
I believe it is a null value. I am doing this request:

`curl -u <username:password> -X POST "http://jenkins.com/job/test-job/buildWithParameters?key1=null"`

I think jenkins converts it to a string, since the parameter in build info JSON for the above build looks like this:

```.
  .
  "actions" : [
    {
      "_class" : "hudson.model.ParametersAction",
      "parameters" : [
        {
          "_class" : "hudson.model.StringParameterValue",
          "name" : "key1",
          "value" : "null"
        }
      ]
    },
.
.
.
```

**Comment by martinda** [comment]
That is not a null value, it is a string with a 4-letter value that spell the word null, but the value is not null. I am looking at [ParametersDefinitionProperty.java](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/ParametersDefinitionProperty.java#L159] trying to see if there is a difference between null and the empty string.

**Comment by richabindra** [comment]
@martinda - that's true. It is a string with the word null. I tried even with the url encoded value for null i.e. "%00", but jenkins build info shows it as an empty string. So, I think there is no easy way to pass null through the rest api. 

I think it's a good idea to convert potential null to an empty list in jenkins-api as discussed earlier. What do you think. @cdancy @martinda

**Comment by richabindra** [comment]
Just saw your comment @martinda . Totally agree with you. I'll go forward with converting null to the empty string. Thanks!

**Comment by cdancy** [comment]
> Just saw your comment @martinda . Totally agree with you. I'll go forward with converting null to an empty string. Thanks!

+1

File: src/test/java/com/cdancy/jenkins/rest/features/JobsApiLiveTest.java [hunk 6]
```diff
@@ -332,6 +332,42 @@ public void testGetBuildCausesOfJob() {
...
+        params.put("SomeKey2", null);
+        IntegerResponse job1 = api.jobsApi().buildWithParameters(null, "JobForEmptyAndNullParams", params);
+        assertNotNull(job1);
+        assertTrue(job1.value() > 0);
+        assertTrue(job1.errors().size() == 0);
+    }
+
+    @Test(dependsOnMethods = "testBuildWithParametersOfJobForEmptyAndNullParams")
+    public void testGetBuildParametersOfJobForEmptyAndNullParams() {
+        while (api().jobInfo(null, "JobForEmptyAndNullParams").lastBuild() == null) {
```
[the remark this answers is not in the document]
Comment (cdancy) [reply src/test/java/com/cdancy/jenkins/rest/features/JobsApiLiveTest.java]: @richabindra if you look at the `QueueApiLiveTest` @martinda added a few tests around queuing builds that is more intuitive at the expense of a bit more code. I would use something like as this could potentially hang forever if someone is not careful. You may have to expose the method `getRunningQueueItem` and potentially make it static.

[the remark this answers is not in the document]
Comment (richabindra) [reply src/test/java/com/cdancy/jenkins/rest/features/JobsApiLiveTest.java]: sure.

[the remark this answers is not in the document]
Comment (martinda) [reply src/test/java/com/cdancy/jenkins/rest/features/JobsApiLiveTest.java]: @richabindra what about moving the method to the `BaseJenkinsApiLiveTest` class?

**Comment by martinda** [comment]
Ah, the live test fails because my admin user is not "admin", it is "martin". When I create the admin user, the test passes. @cdancy is that a concern?

**Comment by richabindra** [comment]
@martinda - ohh.. the test should be user name independent. Should I probably just check if causes list size is greater than 0?

**Comment by martinda** [comment]
@richabindra we're not testing for user name here, so yes the test should be independent of it. Check everything you expect the test to verify, but only that.

**Comment by richabindra** [comment]
@martinda - yea, good catch. Pushed the change!

**Comment by martinda** [comment]
Thank you @richabindra, all the tests pass for me now. This looks good! I have no further comments.

**Comment by cdancy** [comment]
@richabindra merged now. Thanks for the contribution! I'll spin up a new patch release shortly.

**Comment by cdancy** [comment]
Version `0.0.13` with this fix has been released.
